File: python/run-length-encoding/run_length_encoding.py
def prep_encoded_substring(count, char):
    if count == 1:
        return char
    return "{}{}".format(count, char) #char * count

def decode(string):
    solution = []
    for idx in range(len(string)):
        if not string[idx].isdigit() and string[idx-1].isdigit() and string[idx-2].isdigit():
            solution.append(string[idx] * int(string[idx-2] + string[idx-1]))
        if not string[idx].isdigit() and string[idx-1].isdigit() and not string[idx-2].isdigit():
            solution.append(string[idx] * int(string[idx-1]))
        elif not string[idx].isdigit() and not string[idx-1].isdigit():
            solution.append(string[idx])
    return "".join(solution)


def encode(string):
    if not string:
        return ""
    solution = []
    count = 0
    last = ""
    for each in string:
        if each == last or not last:
            last = each
            count += 1
        else:
            solution.append(prep_encoded_substring(count, last))
            last = each
            count = 1
    solution.append(prep_encoded_substring(count, last))
    return "".join(solution)

File: ruby/run-length-encoding/run_length_encoding.rb
# a simple form of data compression, where runs
# (consecutive data elements) are replaced by just one data value and count
class RunLengthEncoding
  def self.encode(str)
    str.chars.chunk_while { |a, b| a == b }
    .map(&:join)
    .collect { |str| str.length > 1 ? str.length.to_s + str.chr : str.chr }
    .join
  end

  def self.decode(str)
    str.scan(/.*?[a-zA-Z ]/)
    .collect { |char| char.length == 1 ? char : char[-1] * char[0..-2].to_i }
    .join
  end
end

module BookKeeping
  VERSION = 3
end
